### DISCUSSION TITLE
Retrieving cost/pricing information?

### DISCUSSION THREAD

**Discussion by ururk** [deepgram/community#240] [General help] [answered]
I looked through the API documentation and didn't see any way to get the total spend on usage. For instance, if I tag my calls:

tag=organization_1

How could I determine how much organization_1 used in terms of money? I see `usd` as part of the result in the API documentation, but when I look over my results there is no `usd` key. I'm also on a trial, so it's possible that isn't sent when the calls are part of free credits.

**Comment by ururk** [comment]
To add to this - I'm really just looking for a summary, not request-by-request details. So that, I could call this endpoint:

https://developers.deepgram.com/reference/summarize-usage

filter and get a USD total along with the summary.

**Comment by jkroll-deepgram** [comment, marked as answer]
Hi @ururk, we have a few options for analyzing cost and usage, but don't currently fully support breaking down cost by tags. You can get your historical spend and hours used by going into your [Project](https://console.deepgram.com/project/) in the console -> Settings -> Billing & Plans. You can also get historical hours for a given tag, either in the Usage Overview in the console, or through the `https://api.deepgram.com/v1/projects/<project-ID>/usage?tag=<tag>` endpoint. Then you can infer the tag's approximate spend based on how many hours you've used for a given tag versus how many hours total you've used. 

You can also get historical requests and their tags with `https://api.deepgram.com/v1/projects/{project_id}/requests`, and see the exact cost of a request with `https://api.deepgram.com/v1/projects/{project_id}/requests/{request_id}`. 

Granted these methods are not ideal, and I will share any updates as we roll out more streamlined methods of cost analysis using tags.

**Reply by ururk** [reply]
Thanks! That is my current plan (we will be using Deepgram internally, and "billing" other users within our org), but it would be nice if it could just spit out a dollar amount for each tag. With how inexpensive `nova` is... getting the number 100% correct isn't critical for our needs.

**Comment by johnkmayo** [comment]
I'm trying to get a download of the individual requests I've made and the cost for each using Python. How do I do that?

**Reply by jjmaldonis** [reply]
Hey @johnkmayo, here is a script you can use as a starting point. This is my WIP script for pulling a range of information about requests:

```python
from typing import Optional, Any
import requests
import json
import os
import datetime
from collections import Counter


DEEPGRAM_API_KEY = os.environ["DEEPGRAM_API_KEY"]  # Your Deepgram API Key

def make_request(url: str) -> bytes:
    headers = {
        "Authorization": f"Token {DEEPGRAM_API_KEY}",
        "content-type": "application/json",
    }
    print(f"Making request: {url}")
    r = requests.get(url=url, headers=headers)
    r.raise_for_status()
    return r.content


def get_project_requests(
    *,
    project_id: str,
    start: Optional[datetime.datetime] = None,
    end: Optional[datetime.datetime] = None,
) -> dict[str, Any]:
    limit = 100
    url = f"https://api.deepgram.com/v1/projects/{project_id}/requests?limit={limit}&page=0"
    if start is not None:
        start = start.replace(microsecond=0)
        start_param = start.isoformat()
        url += f"&start={start_param}"
    if end is not None:
        end = end.replace(microsecond=0)
        end_param = end.isoformat()
        url += f"&end={end_param}"
    result = json.loads(make_request(url))

    # Get all pages
    del result["limit"]
    page = int(result.pop("page"))
    nresults = len(result["requests"])
    while nresults >= limit:
        url = url.replace(f"page={page}", f"page={page+1}")
        _result = json.loads(make_request(url))
        del _result["limit"]
        page = int(_result.pop("page"))
        nresults = len(_result["requests"])
        result["requests"].extend(_result["requests"])
    return result


def summarize_requests(requests: list[dict]):
    print(f"Number of requests: {len(requests)}")

    # Analyse the failed requests
    failed_requests = [
        d for d in requests if "response" not in d or "details" not in d["response"]
    ]
    failed_reasons: Counter[str] = Counter()
    for r in failed_requests:
        failed_reasons[r["response"]["message"]] += 1
    print(f"Number of failed requests: {len(failed_requests)}")
    for message, count in failed_reasons.most_common():
        print(f"{count}: {message}")
        for r in failed_requests:
            if r["response"]["message"] == message:
                print(f"  {r['path']}")

    # Filter out any requests that failed and analyse the successful requests
    requests = [d for d in requests if "response" in d and "details" in d["response"]]
    requests.sort(
        key=lambda d: datetime.datetime.fromisoformat(d["response"]["completed"])
        if "completed" in d["response"]
        else 0
    )
    if len(requests) == 0:
        raise ValueError("All requests failed, cannot summarize.")

    total_seconds = sum(d["response"]["details"]["duration"] for d in requests)
    total_usd = sum(d["response"]["details"]["usd"] for d in requests)

    models: Counter[str] = Counter()
    for d in requests:
        model = d["response"]["details"]["config"]["model"]
        models[model] += 1

    configs: Counter[tuple] = Counter()
    for d in requests:
        config = tuplize(d["response"]["details"]["config"])
        configs[config] += 1

    print()
    print(f"Start: {requests[0]['response']['completed']}")
    print(f"End:   {requests[-1]['response']['completed']}")
    print(f"Number of (successful) requests: {len(requests)}")
    print(f"Total minutes processed: {total_seconds / 60.0}")
    print(f"Total USD: ${total_usd}")
    for model, count in models.most_common():
        print(f"Model `{model}` was used {count} times")
    for config, count in configs.most_common():
        print(f"Config used {count} times: {config}")


def main():
    project_id = "<your project UUID>"

    info = get_project_requests(
        project_id=project_id,
        start=datetime.datetime(year=2023, month=9, day=20),
        end=None,  # datetime.datetime(year=2023, month=9, day=0),
    )
    summarize_requests(info["requests"])


if __name__ == "__main__":
    main()
```

**Reply by johnkmayo** [reply]
Thanks for the starting point (and the super fast response to the question). I had to add an import for datetime and couldn't get the Counter bit working. 

I'm able to get the calls I've made but I'm not seeing the usd element in the response details. So I've got a list of the calls I've made but not what they cost me. Which is progress but just shy of what I need.

Thanks again for the help.

**Reply by jjmaldonis** [reply]
I added those imports to the top of the file. You may need to remove the `: Counter[str]` code depending on your Python version.

All API calls that resulted in a 200 HTTP code should have a `["response"]["details"]["usd"]` field, which is the cost in dollars.

**Reply by johnkmayo** [reply]
I'm using Python 3.10.8 on Windows. I'm getting ["response"]["details"] just not ["usd"] in side of the details.

**Reply by johnkmayo** [reply]
Can we re-open this ticket since I'm getting ["response"]["details"] just not ["usd"] in side of the details and therefore can't track the cost of the individual requests for my project. Or do I need to open a duplicate request?

**Reply by jkroll-deepgram** [reply]
Hi @johnkmayo, can you share a request ID and I'll look into why there isn't a cost associated with it?

Do make sure that you're looking at requests with a 200 response code, since we don't charge for failed requests.

You can also sanity-check in the Deepgram console - https://console.deepgram.com - and go to the Usage tab on the left; Logs; and view past requests or filter by request ID.

**Comment by johnkmayo** [comment]
Request id 9eb99fe12-6032-496d-a433-9e5c03ab8f4c

**Reply by jkroll-deepgram** [reply]
Could you double-check that request ID? The first section (`9eb99fe12`) has 9 characters, but it should have 8 characters.

**Comment by johnkmayo** [comment]
Sorry about that. It is 9eb9fe12-6032-496d-a433-9e5c03ab8f4c. I had to
rekey it and got it wrong. Thanks for looking into this.

Here are a few others in case that helps:
4928a35c-9866-450b-8c15-e86d8664059b
fd75fbe2-4c13-431e-966d-05e66653af5d
5b763b88-4b48-44df-b4a8-55c1e3cc182f
055ed799-b28b-46f1-9b1f-ff0b3580919e
8f097a18-d6bd-4309-b453-ba4d038fa839
48a8f66e-0b68-4298-a106-e15143415ad1
e51d76da-83e9-4787-b82a-5e60a2c6038d
3de3af32-3654-44ab-8058-b73310f64a24
68b58525-f2ff-4b88-bf8f-c8d5cab374e3
1c28e1f1-3850-4b3c-8e12-d368504af26f
bfed3609-18fe-4190-946d-37cafe9bac69
39218d39-2537-45e8-a600-2d1a1a19adf3
ff2da676-e547-42f4-9d7f-f434451dc1ff
65fbe3c1-b81c-4ece-82ca-a950fe612439
0fc89390-fc22-4015-bc15-75b898588cc1
b8d03be0-5867-4059-996e-8e2247fd3974
d52e4d2e-2140-47bd-9acb-ead0529417ca
fa2d1b01-aceb-42dc-adf6-2e57806cd40e
3e2afe7f-dc01-4ce3-8c6f-b7fc578aacf2
06176dd0-5c0c-4c83-88d3-4b563245190d
6bb28b40-5228-45a2-9086-926c496054dd
2451d4e3-a300-43bc-9255-5fc93298308c
f53847d0-113c-42be-8523-62183c5dff2d
b68cf02a-0678-43e3-8add-8d3a41aa9453
1da13df2-8e6d-47ec-8105-af7b87d43c19
d907c73a-cb16-48d9-a006-31e1474600d1
1d6cbf66-706e-420c-a51c-86193c12a020
61a59367-5045-4188-b261-3a345a200166
f5739a75-cbd0-4b55-b29b-667ab767fc29
676b3798-e698-4d43-9a47-7669c6931bbc
8c4baba9-548b-4c51-94c1-34bb4d373ab8
b97748e5-ff3b-468f-a0ba-8938e0a04f66
f00da654-354c-4448-8e0e-37aab23dcaa9
863d9809-685f-4b70-99de-b4a3d1a3fed2
8348bad2-8318-475f-8ee9-d7d6cd9e929d
e70082ca-96f4-474b-ac02-7a8414f42be1
42a79c06-c7ad-477d-80a4-b8cffb89daae
2c9e29fe-cb6f-4c8c-8cc4-d8e3f3f4eef4
812564b9-3e94-4344-8f49-d9488c9b0a2f
fac27fcf-6353-4aaa-9f13-267f1207b382
ae29a6d6-1e02-4dde-a863-725ba7af539c
08527dff-aa31-4ceb-8590-0de9f7a843cf
495d0491-7d92-424b-b999-40dd647f9082
e89df515-aa79-402e-9d79-64a540d6ebfe
7bf31472-a4ad-4139-a47e-93ae6635a938
93de4af4-b34e-4fb2-b5f0-4213fd3c39d6
0c4c8092-3abc-4f6d-9b18-7bdbf4c4be17
a608f06a-3877-4558-b963-eba63c9e7504
0adbe931-3733-4656-a1c9-23bc77ab4061
1f801442-0c4d-4d64-9ea7-a0acd1496af9
af025191-2011-4d4c-ab5d-ec1ba0649348
39100f98-6592-49f6-93fd-7246411c7190
31686487-cfdf-4a51-954d-18cef699ee02
2b6acbf1-00ad-4632-b7d8-c9871c1b70d2
d658127b-ee5e-41a1-bd8d-d7a694b8f13c
a90c511d-92ec-4ad0-9ab4-88cb97c9b739
05c87e36-beaa-44a3-ba6f-1880289730ab
8e37aef2-d719-4930-b670-6ff62454d891
97da3007-f140-44c2-886a-3c67fa2c2d7f
f1e55e86-bc5e-49c7-9127-5ae93c26a36a
ae790631-5630-4b0d-95ee-640d84f5e075
d180cdb4-5912-4a7c-8c5f-06b822be5c73
afaa7000-78c0-4f01-a59f-50047959be1a
bfdb8930-6a46-4ff4-996e-51870287374b
8c8dd924-95b8-477e-b245-e756bcf3dfa8
f091749d-5516-420a-aaf2-099dbd8ffaef
a35448d4-db2c-4a30-977c-444437aed095
040fefb8-f1b6-4370-be60-88320c4546cd
1539b2e7-f223-4a73-baf4-fe6e324925d7
41072b42-ca91-4b37-b832-cee9519d37e8
97696104-6ff1-4184-9502-c28f84c7c18e
bbc1ea2d-ee37-40e1-9b44-099defd65d05
60facc53-0265-4bee-83c3-7df81df6e33c
ead02213-3179-4d32-9c02-7af0c3942260
18ac7b46-dbb2-4074-b9f2-bbfa6973c554
da149b94-0bc6-438c-a504-84247a2283d1
345f85b3-0426-4263-a308-86121ec4a1cd
70e4ce47-86c3-415a-ab39-88dd9152b05b
96d9ee3d-3f5a-4329-bd8f-2308d6eba6a2
3789a4c9-b378-429d-98a8-0e07a529a0bb
0b7ff6f0-6e7c-4d31-a2ce-41031f2170b9
77479bc9-f57a-4fff-a05b-07644a3d1eb8
5703ee44-0e25-4014-bdd4-78b62a03a928
81fd9cbd-4101-4845-b428-8d8aac743af1
16cae40a-b283-4e34-92e0-741ec47cbdd9
e69a5e6a-8e33-4c9f-9fb3-0041e885de4e
b9047f58-f55c-48ae-8e63-bff3e387f612
2ce8ee40-9f86-495e-8ce3-4cace6547cba
918f9eed-162a-48ca-a189-7e720c3a7f11
437b2ac9-f089-452e-97f1-3f8478d6e3bf
1a2ff6be-2631-4783-9427-1d295af8f629
0cf635f7-1b26-449a-bbb9-5863b5bc40c9
91022e28-879d-45e6-86d2-d4aea81ae90c
16d2001f-144e-48bb-b02f-5c00b26f2014
098ec42f-ed16-4521-a154-951e0835d8db
78b68ea3-0c75-4ff4-b939-1b9f80c7804f
3aaae78a-39b9-43e0-af0d-7b1b776ca2f8
6f7993a8-7e97-42a0-98ec-fe4cca8b0127
063d1c25-358d-4960-8ea5-91b03c4e35f1
63a0d383-fee8-42bc-b8d4-dd4601a01cf1
2cb5c27a-9ab2-460e-a7dd-a6bdcb60a0c6
ae452e3d-3d17-4a3c-9f21-b25e2d762cd2
4d9a344e-29e6-4d2a-b95e-a7a7da2819ed
4e362592-136b-4c79-a76c-50436bdb5852
e7bdf67a-0204-4ada-9dc5-18d25071d69e
04b00ff8-0f37-4281-adf0-38b1f12c47b4
43f60a7c-7f16-44d8-8f2a-21e4ae6cfea4
2aa7e273-737f-42ee-95a6-7e2e9e8ebcbb
d1db68eb-e0ec-4419-8fd3-0f6576750e1e
05e405fa-42e7-470e-96a8-f1a0044c458d
d96be72d-e456-4ec0-a0a3-29ca7f29c62a
b31cbe9f-a484-4f4d-a4f6-df7537b62853
dd320cde-08ae-494b-b9b9-d519b1f909b4
97092f92-f661-4030-82f1-88c7091f96e8
52147b6d-70c6-4f84-aec2-5fa06d8dbe2f
ded18af3-40de-4d0f-a474-d66ccc6e63fc
51e701bc-9a7a-4629-8b51-9a25d0918e97
50eadb8e-d915-4b9c-8c5d-fb131dbb84c1
340282d9-7fcb-4a21-aed1-ef339dc42eab
a3172a25-b0fa-45d1-a2b2-d9cca81a631c
3e977f9d-2fe1-46b3-bc13-2d80d207cee7
0832421e-9731-4708-ac6d-bdb8a95e5c5d
77eca4c6-01ef-475a-a54d-f6b7efb56491
c852e876-5c29-4629-8c00-375a3fa60857
d2dcf849-18ef-4cd3-9981-ecfa5cbf9aef
1ab176ea-1bed-4514-a67b-41871e9b6760
a60a41da-78a4-4254-a6ae-22507da9b976
87c75fe9-034c-435f-89c4-6a60743b9aa1
866b88d3-ccf1-4fca-8e8b-22a02e689f8a
7b677690-111c-489f-87d7-94db8a97d999
bf92909d-70ec-44b6-9138-0aba4e42d8d3
ae86e7e9-ff52-4a30-a3b2-863ec25952ff
045d2d9a-0de2-46d7-a954-74cc1da5c3fb
8ddfcfc9-2e9f-4451-96fb-7ebfd26a94e1
18ba83d5-7347-403e-9fed-59e1455d0c32
11f57612-fca8-4b3a-99f7-a6b87d6c76ee
d3babac8-ad23-490a-99e4-80a0559b47e3
23fa3198-8909-4426-adf7-8078499c8b6b
445423bd-1ba5-4d9e-b931-b0b4e83476b9
feb2292b-c0dc-4e26-a12f-8c2ef44acf1f
0227fa0a-fb74-4537-a34a-98d145d5b3f7
ab299b57-3769-4373-bb52-ed4813511336
6c239873-a2dc-4eb0-80a3-88426dcc9abe
e7442507-9905-409e-a54a-adf1e859e0f1
81f00cb0-9448-431f-a58b-b31ee2b1ebc6
7dd0d440-cc35-495c-8dfb-1f14899785fa
057a4ad2-4f75-45a5-9ea7-319d5d8c65cd
07413763-2546-4e9d-8564-96f6c4762570
b65580dc-5653-44f5-8c01-3b7ea830b677
7154a217-50c1-4fcc-bdb7-d51288fd7a11
72c89122-c8a7-4100-be84-a0e833a660df
3bc811bc-2ebf-426b-8f7b-9d036acd6633

**Reply by jkroll-deepgram** [reply]
Thanks for providing these request IDs! It's looking like this has to do with your API key permissions. API keys with a "member" role are not allowed to access billing information. If you create a new API key with an "admin" or "owner" role, then your same API calls using that new key will include the cost for each request. Check out our [roles and scopes docs](https://developers.deepgram.com/docs/working-with-roles) as well.

**Comment by johnkmayo** [comment]
That worked. Thanks!

It is a shame the filename or URL isn't returned in the results. That would
make it much easier to match the cost to the file transcribed.

**Reply by jkroll-deepgram** [reply]
Glad to hear that the new API key worked for you!

Thanks for the feedback on working with results. Consider that you could submit the same file or URL for multiple requests - it's not guaranteed to be unique. I recommend using request ID to map from each request to response, as it will always be unique and is intended for that purpose.

**Comment by johnkmayo** [comment]
Got it. I just didn't think to record the request ids the other day. I will
try to add that to my process. Thanks.
